src/sounds.cpp

```diff
@@ -332,15 +332,23 @@ void sounds::process_sound_markers( player *p )
 
         if( !p->has_effect( effect_sleep ) && p->has_effect( effect_alarm_clock ) &&
             !p->has_bionic( bionic_id( "bio_watch" ) ) ) {
+            // if we don't have effect_sleep but we're in_sleep_state, either
+            // we were trying to fall asleep for so long our alarm is now going
+            // off or something disturbed us while trying to sleep
+            const bool trying_to_sleep = p->in_sleep_state();
             if( p->get_effect( effect_alarm_clock ).get_duration() == 1_turns ) {
                 if( slept_through ) {
                     p->add_msg_if_player( _( "Your alarm-clock finally wakes you up." ) );
-                } else {
+                } else if( !trying_to_sleep ) {
                     p->add_msg_if_player( _( "Your alarm-clock wakes you up." ) );
+                } else {
+                    p->add_msg_if_player( _( "Your alarm-clock went off and you haven't slept a wink." ) );
                 }
                 p->add_msg_if_player( _( "You turn off your alarm-clock." ) );
             }
-            p->get_effect( effect_alarm_clock ).set_duration( 0_turns );
+            if( !trying_to_sleep ) {
+                p->get_effect( effect_alarm_clock ).set_duration( 0_turns );
+            }
```

Comment:
I'm not sure if this line is correct. If the player wakes up to some noise not coming from the alarm (like a zombie knocking at the door), `trying_to_sleep` will be true, and the alarm clock effect will not be cleared, which might cause problems if they try to sleep for a second time without setting the alarm.

Reply:
Looking at this again I think you're correct.  I'm not certain why I thought checking for trying_for_sleep was necessary.  Will fix.  Thanks!